challenger.go

```diff
@@ -79,3 +214,39 @@ func (l *LndChallenger) NewChallenge(price int64) (string, lntypes.Hash, error)
 
 	return response.PaymentRequest, paymentHash, nil
 }
+
+// VerifyInvoiceStatus checks that an invoice identified by a payment
+// hash has the desired status. To make sure we don't fail while the
+// invoice update is still on its way, we try several times until either
+// the desired status is set or the given timeout is reached.
+//
+// NOTE: This is part of the auth.InvoiceChecker interface.
+func (l *LndChallenger) VerifyInvoiceStatus(hash lntypes.Hash,
+	state lnrpc.Invoice_InvoiceState, timeout time.Duration) error {
+
+	exitTimer := time.After(timeout)
+	for {
+		l.invoicesMtx.Lock()
```

Comment:
If we want to avoid spinning here (even with the interval) we can use another map to store a series of [condition variables](https://golang.org/pkg/sync/#Cond), which will be signaled once the entry is populated. A rough sketch would be something like:
  * grab lock
  * check if invoice is there
  * if so, gg
  * if not, add a new condition variable to the signal map (specific for this invoice) 
  * wait until signaled 

Alternatively, we can just use a single condition variable for (invoice state has been updated), which'll let us avoid needing to tack on yet another map. This route is simpler and lets us avoid polling the map, making everything a bit more async.

Reply:
Very nice idea! I've never used the `sync.Cond` before, so I went with the suggestion. The code _looks_ a bit more complicated but I think it's not too bad. And there certainly isn't any spinning involved anymore.